### PR TITLE
Omer links

### PR DESCRIPTION
**Pull request by omer88**

### CODE REVIEW

File: docs/installation-android.md [hunk 1]
```diff
@@ -206,3 +209,20 @@ dependencies {
   compile "com.google.firebase:firebase-perf:11.4.2"
 }
 ```
+
+## 7) Dynamic Links (optional)
+
+If you plan on using [Firebase Dynamic
+Links](https://firebase.google.com/docs/dynamic-links/):
+
+In `android/app/src/main/AndroidManifest.xml`, add a new intent filter to the activity that handles deep links for your app, and specify the host and the scheme:
```
Comment (erezrokah) [inline docs/installation-android.md]: You can add "for react-native this is usually `MainActivity`"

File: docs/installation-android.md [hunk 2]
```diff
@@ -206,3 +209,20 @@ dependencies {
   compile "com.google.firebase:firebase-perf:11.4.2"
 }
 ```
+
+## 7) Dynamic Links (optional)
```
Comment (erezrokah) [inline docs/installation-android.md:213]: You should reference `https://firebase.google.com/docs/dynamic-links/android/receive`.
Something like:
"Make sure to setup dynamic links for Android as described here..."

People need to know that they need to enable this feature

File: docs/installation-ios.md [hunk 3]
```diff
@@ -159,3 +159,68 @@ didReceiveNotificationResponse:(UNNotificationResponse *)response
 If you're having problems with messages not being received, check out the following blog post for help:
 
 https://firebase.googleblog.com/2017/01/debugging-firebase-cloud-messaging-on.html
+
+
+
+## 5) Dynamic Links (optional)
```
Comment (erezrokah) [inline docs/installation-ios.md:165]: You should reference `https://firebase.google.com/docs/dynamic-links/ios/receive`.
Something like:
"Make sure to setup dynamic links for iOS as described here..."

File: docs/installation-ios.md [hunk 4]
```diff
@@ -159,3 +159,68 @@ didReceiveNotificationResponse:(UNNotificationResponse *)response
...
+
+If you plan on using [Firebase Dynamic Links](https://firebase.google.com/docs/dynamic-links/) then, you need to:
+
+### 5.1)  create a new URL type
```
Comment (erezrokah) [inline docs/installation-ios.md]: I would consider adding screen shots for `5.1` and `5.2` as it is not noobs friendly (I know it is taken from firebase docs, but remember how we got lost in Xcode settings)

File: docs/modules/links.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
+# Dynamic Links
+
+[Firebase Dynamic Links](https://firebase.google.com/docs/dynamic-links/) allows you to create and receive links on both Android and iOS platforms. Assuming the installation instructions have been followed, Firebase Dynamic Links is ready to go.
+
+
+All Dynamic Links operations are accessed via `firebase.links()`
+
+## Create Dynamic Links
+
+RNFirebase mimics [The REST API](https://firebase.google.com/docs/dynamic-links/rest) for Dynamic Links creation.
```
Comment (sagi) [inline docs/modules/links.md]: RNFirebase mimics [Firebase's REST API](https://) for ...

File: docs/modules/links.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+#### `onLink(listener: Function): () => any`
+
+On a new URL, the payload URL is passed to the listener callback. This method is only triggered when the app is running. Use getInitialLink for URLs which cause the app to open.
+In order to subscribe to the listener, call to the method with a callback and save the returned function.
+When you want to unsubscribe, just call the function that returned at subscription.
+
+```javascript
+// Subscribe
+const unsubscribe = firebase.links().onLink().then((url) => {
```
Comment (erezrokah) [inline docs/modules/links.md]: This should be:
```
const unsubscribe = firebase.links().onLink(url => {
  //...
});

File: docs/modules/links.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+Call getInitialLink to access the URL that the app has been launched from. If the app was not launched from a URL, the return value is null.
+
+```javascript
+firebase.links().getInitialLink().then((url) => {
+  //...
+});
+```
+
+#### `onLink(listener: Function): () => any`
```
Comment (erezrokah) [inline docs/modules/links.md]: Should we change this to:
```
onLink(listener: Function<string>): () => void
```

File: docs/modules/links.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+```
+?> Please note that when using `androidInfo` or `iosInfo`, `androidPackageName` and `iosBundleId` are mandatory (respectively).
+
+For more information [see reference](https://firebase.google.com/docs/reference/dynamic-links/link-shortener)
+
+## Receive Dynamic Links
+
+### Methods
+
+#### `getInitialLink(): Promise<String>`
```
Comment (erezrokah) [inline docs/modules/links.md:101]: It looks like they use lowercase `string` in their docs:
https://github.com/rmrs/react-native-firebase/blob/omer_links/docs/modules/cloud-messaging.md

File: tests/src/tests/links/linksTests.js [hunk 9]
```diff
@@ -0,0 +1,441 @@
...
+            socialDescription,
+            socialImageLink,
+          },
+        },
+      };
+
+      const result = await links.createShortDynamicLink(data);
+      result.should.startWith(`https://${dynamicLinkDomain}`);
+
+      const response = await fetch(result);
```
Comment (erezrokah) [inline tests/src/tests/links/linksTests.js:380]: This introduces a linter error. You should add `fetch` as a global under `.eslintrc`

File: tests/src/tests/links/linksTests.js [hunk 10]
```diff
@@ -0,0 +1,441 @@
...
+        },
+        suffix: {
+          option: 'SHORT',
+        },
+      };
+
+      const result = await links.createShortDynamicLink(data);
+      result.should.startWith(`https://${dynamicLinkDomain}`);
+
+      const response = await fetch(result);
```
Comment (erezrokah) [inline tests/src/tests/links/linksTests.js:417]: See above comment about `fetch`

File: tests/src/tests/links/linksTests.js [hunk 11]
```diff
@@ -0,0 +1,441 @@
...
+    });
+
+    it('getInitialLink should return null', async () => {
+      const initialLink = await links.getInitialLink();
+      should(initialLink).be.null();
+      Promise.resolve();
+    });
+
+    it('should listen to link', () => {
+      const unsubscribe = links.onLink((url: string) => {
```
Comment (erezrokah) [inline tests/src/tests/links/linksTests.js]: this is also a linter error. You should add `console.log(url);`

File: docs/modules/links.md [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+      socialDescription: string,
+      socialImageLink: string
+    }
+  },
+  suffix: {
+    option: "SHORT" or "UNGUESSABLE"
+  }
+}
+```
+?> Please note that when using `androidInfo` or `iosInfo`, `androidPackageName` and `iosBundleId` are mandatory (respectively).
```
Comment (erezrokah) [inline docs/modules/links.md:93]: `?> ` prefix here is on purpose?

File: docs/modules/links.md (hunk 6 above)
Comment (omer88) [reply docs/modules/links.md]: oops.

File: docs/modules/links.md (hunk 7 above)
Comment (omer88) [reply docs/modules/links.md]: I don't know, it's also the same for messages, since the function is returned from react's event emitter.

Comment (erezrokah) [reply docs/modules/links.md]: `onMessage()` doesn't specify a return value in the docs...
The argument for `onMessage()` is `listener: Function<Object>`

File: docs/modules/links.md (hunk 8 above)
Comment (omer88) [reply docs/modules/links.md:101]: But on other palaces in the code they used String (for example lib/modules/auth/index.js)

**Comment by erezrokah** [comment]
There is one `core` test failing but I don't think it is related to you. Looks like this:
https://github.com/invertase/react-native-firebase/commit/4f5fdda1f0860e82dacba8949e4162ffa8366f29#diff-37dfb27cca404b75acb13ae8afbceca3 
broke it

File: docs/modules/links.md (hunk 12 above)
Comment (omer88) [reply docs/modules/links.md:93]: Yes. see: https://github.com/rmrs/react-native-firebase/blob/omer_links/docs/modules/firestore.md vs. https://invertase.io/react-native-firebase/#/modules/firestore

**Comment by erezrokah** [comment]
`npm run internal-tests` also fails (3 tests), but it also fails on `master`

**Comment by erezrokah** [comment]
If you run ` npm run docs-serve-local` `Dynamic links` is missing from the side bar.
You should run `npm i docsify-cli -g` first.
It looks `docs/_sidebar.md` is missing a reference to dynamics links.

**Comment by sagi** [comment]
**docsify-cli** looks awesome

**Comment by erezrokah** [comment]
I think the docs should be auto generated some how from the code comments (still trying to figure this out).
I couldn't find where (if) they auto generate the docs, @omer88 you should verify with Chris.

File: lib/modules/links/index.js [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    super(firebaseApp, options, true);
+  }
+
+  get EVENT_TYPE() {
+    return EVENT_TYPE;
+  }
+
+  /**
+   * Returns the link that triggered application open
+   * @returns {*}
```
Comment (erezrokah) [inline lib/modules/links/index.js]: Should return `@returns {*|Promise.<String>}` see `messaging().getToken()`

File: lib/modules/links/index.js [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+   */
+  getInitialLink() {
+    return this._native.getInitialLink();
+  }
+
+  /**
+   * Subscribe to dynamic links
+   * @param listener
+   * @returns {*}
```
Comment (erezrokah) [inline lib/modules/links/index.js]: Should return `Function`?

File: lib/modules/links/index.js [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+   */
+  onLink(listener: Function): () => any {
+    const rnListener = this._eventEmitter.addListener(EVENT_TYPE.Link, listener);
+    return () => rnListener.remove();
+  }
+
+  createDynamicLink(parameters: Object = {}): Promise<String> {
```
Comment (erezrokah) [inline lib/modules/links/index.js]: Missing comments

File: lib/modules/links/index.js [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+    return this._native.createDynamicLink(parameters);
+  }
+
+  createShortDynamicLink(parameters: Object = {}): Promise<String> {
```
Comment (erezrokah) [inline lib/modules/links/index.js]: Missing comments

File: android/src/main/java/io/invertase/firebase/links/RNFirebaseLinks.java [hunk 17]
```diff
@@ -0,0 +1,332 @@
...
+      if (iosParameters.containsKey("iosIpadFallbackLink")) {
+        iosParametersBuilder.setIpadFallbackUrl(Uri.parse((String) iosParameters.get("iosIpadFallbackLink")));
+        iosParameters.remove("iosIpadFallbackLink");
+      }
+      if (iosParameters.containsKey("iosMinPackageVersionCode")) {
+        iosParametersBuilder.setMinimumVersion((String) iosParameters.get("iosMinPackageVersionCode"));
+        iosParameters.remove("iosMinPackageVersionCode");
+      }
+
+      if (iosParameters.size() > 0) {
```
Comment (erezrokah) [inline android/src/main/java/io/invertase/firebase/links/RNFirebaseLinks.java]: This validation code is duplicated with iOS.
See `messaging().scheduleLocalNotification()` where they validate the object in JavaScript.
I'm not sure what is more efficient (but I think we should avoid sending non relevant data over the bridge and avoid code duplication)

**Comment by erezrokah** [comment]
The validation code for `parameters` is duplicated between Android and iOS.
See messaging().scheduleLocalNotification() where they validate the object in JavaScript.
I'm not sure what is more efficient (but I think we should avoid sending non relevant data over the bridge and avoid code duplication)

File: docs/modules/links.md [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+
+#### `createDynamicLink(parameters: Object): Promise<String>`
+
+Creates a long dynamic link.
+
+```javascript
+firebase.links().createDynamicLink({
+  dynamicLinkInfo: {
+    dynamicLinkDomain: "abc123.app.goo.gl",
+    link: "https://example.com/",
```
Comment (erezrokah) [inline docs/modules/links.md]: I would change `link: "https://example.com/",` to `link: "https://example.com?param1=foo&param2=bar",` so it is clear you can specify parameters

**Comment by erezrokah** [comment]
@omer88 see my change here:
https://github.com/rmrs/react-native-firebase/commit/64ba93936106cad62ae1d4ef9d24d30a2f16df1e

I think we can use the same method for the other tests. WDYT?

File: lib/modules/links/index.js (hunk 13 above)
Comment (omer88) [reply lib/modules/links/index.js]: why not only `Promise.<String>`?

Comment (erezrokah) [reply lib/modules/links/index.js]: If you can be explicit about the type I think it is better

Comment (omer88) [reply lib/modules/links/index.js]: I don't see a reason why not. a promise is returned and in the worst case we reject it or the promise value is null.

**Pull request closed without merging**